### ISSUE TITLE
Proposal: Update labels 

### ISSUE THREAD

**Issue by vdemeester** [docker/docker#21721]
This is a follow-up issue to the closed PR #18958 that was only for containers.

Now that networks, volumes, images and containers all support labels, we should probably rethink the idea (or not) of having a common way to update them, instead of only being able to update containers label and no others.

Passing @tiborvass comment from https://github.com/docker/docker/issues/20356#issuecomment-194467286

/cc @icecrime @tiborvass @calavera @dnephin @ehazlett 

---

Before going into technical details, I would like to find a good UX. I could come up with these 3 solutions, none of which I really like, but I prefer the 3rd one among these:

1. For each new object type (images, networks, volumes), add a new `update` subcommand to match the `docker update` that is for containers. Questions: is there anything else we would want to update on images networks and volumes? If so what? If we can't agree on those other things to update, then the problem with this solution is we would add a new subcommand just for labels. Moreover, for historical reasons `images` is not a subcommand the way `network` and `volume` are. So if we were to still go with this solution we should also choose between: 1.a) introduce `docker images update` subcommand OR 1.b) overload `docker update` the way we (unfortunately) overload `docker inspect` and accept: `docker update --label foo=bar $imageID`.

2. Overload `docker update` to non-container objects as well. Problem: how to deal with `docker update --memory 2G $networkID` ? obviously it should error out not recognizing the memory flag for updating network objects, but we would have the same problem docker machine is having with all the cloud-provider-specific flags and it's not ideal for the help output. Also, if we are overloading `docker update` for non-container objects, why not `docker create` and more ? Since this solution is a generalization of 1.b, 1.b also suffers of this problem. Both solutions 1 and 2 would result in accepting #18958.

3. Introduce `docker label` command that can act on any object. Problem: we already have `docker update` that updates the container settings (cgroups only for now) and this would be another toplevel command that would also update container settings, albeit only one setting: the labels. As much as I dislike having two commands do similar things, a counter-argument is to say: labels are special because they are meta. Obviously this would mean, the UX in #18958 would be rejected.

If you can think of another solution that would have none of these UX issues, I would be glad to read about it.

**Comment by bboreham** [comment]
> is there anything else we would want to update on [...] networks

Of all the current options to `docker network create`, `--ip-range` is the main one that seems amenable to changing after the network is in use.  And then it would require that the particular IPAM driver is implemented in such a way that its range can be changed.

`--opt`, which is a catch-all set of options for the driver, is another possibility.  Though one might consider replacing this mechanism with labels.

**Comment by duglin** [comment]
I'd prefer a top-level "docker image" command and then we can do:
```
docker image update ...
docker container update ...
docker volume update ...
...
```
To me, a label isn't a top-level object so having "docker label ..." feels odd.  Its a property of some other object.

https://github.com/docker/docker/issues/13509

**Comment by thaJeztah** [comment]
@duglin's suggestion would be my preferred solution, but obviously the one with most impact as well

**Comment by thaJeztah** [comment]
Having mutable container (runtime) labels recently came up in a discussion
on the internal Slack channel. A short summary of that discussion;



It's currently not possible to change labels on a running container. In
addition, containers inherit labels from the image they're run from, and any
label set during "docker run", is appended to that list.

- containers should have mutable (runtime) labels, i.e. I should be able to set
  a `active=true` label on a container. n.b. UX for setting/unsetting/updating
  labels was not discussed, but should be consistent for all resources, re
  https://github.com/docker/docker/issues/20356#issuecomment-194467286
- container (runtime) and image labels should be separate properties, and not
  be merged. (note: currently, `docker run --label foo=bar` merges the label with
  labels that are present in the image - so this should change)
- to preserve backward-compatibility, we can use both "image" and "container"
  labels when filtering, i.e. --filter foo=bar would use both. (Q: in case of
  conflict, which label takes precedence?). It might be easier to give a
  deprecation warning then fix the filter (implementing `--filter
  image-label=xyz=123`). Worst case if someone misses the deprecation warning
  they can add the "image" prefix to their code and it will work again
- to filter containers on by the labels of their image, we can add a
  `--filter image-label=xyz=123` (see previous bullet)

/cc @shykes

**Comment by aanm** [comment]
@thaJeztah I also think it would be nice to disable labels from a particular image when starting a container.

**Comment by creynders** [comment]
Mutable container labels would be _really_ handy. I'm using [traefik](https://traefik.io/), a reverse proxy, which automatically detects docker containers and reads traefik-namespaced container labels for automatic configuration. Setting a label on a running container would allow on-the-fly rule changes.

**Comment by blabno** [comment]
I'm pooling containers that take long time to start and are used later by other containers as dependencies. It would be awesome to mark container as "used" when it gets linked.
Right now I have to maintain map of all containers and their state externally.

**Comment by yarmiganosca** [comment]
In our deployment pipeline, we do blue-green deploys with containers. We launch N new containers, & wait for them to register with consul as healthy before stopping & removing the old containers. If the new containers register as unhealthy, or don't register at all, before our timeout, we stop & remove the new containers. What we'd like to do in that case is keep one of the failed/unhealthy containers around so we can inspect & troubleshoot--a dead canary, if you will. Automating the cleanup of any such dead canary containers on the next deploy gets a lot easier if we can relabel running containers.

**Comment by danbeaulieu** [comment]
Similar use case to other users where we pool containers as warm standbys. Once we activate them we'd like to label them for accounting/monitoring etc systems which are part of systems decoupled from the activation system.

**Comment by whytoe** [comment]
Docker Cloudstor:Azure plugin uses a unique identifier when creating volumes in Azure
![screenshot 2017-07-16 at 11 39 30](https://user-images.githubusercontent.com/12265878/28249039-b9cac5be-6a1c-11e7-821b-e8af1c0f6b2c.png) and it does not input the actual Volume name.

Updating a label to correlate the Docker Volume to the Azure "Container" would be very helpful

**Comment by sjeeva** [comment]
Is label update for containers supported? I couldn't find any options in **docker container update** command

**Comment by jshnaidman** [comment]
Without being able to update labels, there's currently no way to update any metadata or description or any fields pertaining to a container. Very inconvenient. Would be very useful to be able to update labels or some other custom field.

**Comment by g4gupta** [comment]
whats the status here ... am looking for a way to be able to update labels of my images . I would like to manage my images and metadata to them based on their last usage . As such the contents of the image will not change hence I don't want to build a new image for it to just add /delete some metedata info pertaining to that image .

**Comment by aJetHorn** [comment]
One use case is extremely time-sensitive hotfixes that are done by using `docker exec` and `docker restart`. Let's assume it's too time-consuming to send a change through any CD pipeline or even rebuild the container (I understand that in _most_ cases, especially ones using lightweight images this difference is negligible and this approach is unnecessarily dangerous). Once a hotfix is complete, I will use `docker commit` to create a new image with the proper *-hotfix-* tag (, do whatever I need to do git-wise) and then upload it to my registry. At that point, I'm guaranteed that the repo image and the running image are running the same code, yet the tag of the running image can't be updated. Being able to affix a "HOTFIX" label before I am able to re-run the image I've committed from the hotfixed container would be helpful. For the sake of example, assume the container in question isn't run redundantly but is high traffic and can only be restarted during one hour of downtime during the day.

**Comment by shadowbq** [comment]
Again.. another usecase. v2tec\watchtower will update docker images based on LABELS. If I can dynamically add/remove the labels I can control the automatic update cycle in the case of an upstream bad commit/failing docker image version.

**Comment by sooryamanian91** [comment]
Another use case, when updating the labels for K8S PODS using kubectl, the updated POD label would not reflect in the docker container of the POD.

**Comment by G1itcher** [comment]
Seeing how many docker tooling images seem to have gravitated towards using labels to define their interactions (watchtower, traefik, etc), it's becoming increasingly important to allow labels to be changed. For example, I have a server with 44 images on them and am looking to move my reverse proxy over the traefik. I now have to go through each image (luckily I made docker-compose files for most of them) and add the appropriate labels and re-up them. Even assuming they all come back okay, that's a big chunk of work

**Comment by shenshouer** [comment]
Any plan for this ?

**Comment by lucwillems** [comment]
seems swarm service has this option : --container-label-add --container-label-rm
any reason why we can't do this on a "plain" container

**Comment by lovejoy** [comment]
if update a pod's label in k8s, the label of the pause container not update, seems related to this issue ?

**Comment by huzhengchuan** [comment]
Any plan for this ?

**Comment by slavanap** [comment]
(May be related) Is the a way to hide some running containers and image tags for them from showing at `docker ps -a` and `docker images` and that also be immune to `docker system prune -a`? Dockerfiles testing and development becomes much slower when you have some services running in Docker while developing Dockerfiles at the same PC...

**Comment by juanjoDiaz** [comment]
I'm still looking for this functionality.
Any progress on this?

**Comment by tvon** [comment]
If it is of any use to anyone, I wrote [a script to add labels to a Docker tarball](https://gist.github.com/tvon/edeaf08b5e68c651b89a7e5a7e4dbfd9):

```bash
#!/bin/bash
#
# Add a label to an image tarball (manifest v1).
#
# $ docker save alpine -o alpine.tar
# $ add-label alpine.tar org.mine.key foobar
# $ docker load -i alpine.tar
# $ docker inspect alpine | jq '.[].Config.Labels'
# {
#   "org.mine.key": "foobar"
# }

set -euf -o pipefail

tarball=${1:-''}
key=${2:-''}
val=${3:-''}

if [[ -z "${tarball}" || -z "${key}" || -z "${val}" ]]; then
  echo "Usage: add-label image.tar key value" && exit 1
fi

workdir=$(mktemp -d)
imagedir=${workdir}/image
mkdir "${imagedir}"

cleanup(){
  rc=${1:-0}
  rm -rf "${workdir}"
  exit "${rc}"
}

trap cleanup EXIT

tar -C "${imagedir}" -xf "${tarball}"

# Figure out which file is the config json.
config=$(jq -r '.[].Config' "${imagedir}/manifest.json")

# modify config.Labels to a tempfile
jq -c \
  --arg key "${key}" \
  --arg val "${val}" \
  '. * {config: { Labels: { ($key): $val }}}' "${imagedir}/${config}" \
  > "${workdir}/config.json"

# modify container_config.Labels and overwrite original file.
jq -c \
  --arg key "${key}" \
  --arg val "${val}" \
  '. * {container_config: { Labels: { ($key): $val }}}'  "${workdir}/config.json" \
  > "${imagedir}/${config}"

mv -f "${tarball}" "${tarball}.orig"
pushd "${imagedir}" > /dev/null

# shellcheck disable=SC2046
tar -cf "${workdir}/out.tar" $(ls)

popd > /dev/null

mv "${workdir}/out.tar" "${tarball}"
```
I'm not sure if I should be fiddling with both `config.Labels` and `container_config.Labels`, but it appears to work.

If it breaks then you get to keep both pieces.

**Comment by bluebrown** [comment]
@lucwillems 

> seems swarm service has this option : --container-label-add --container-label-rm
any reason why we can't do this on a "plain" container

This will cause the container to restart. So you can do this in plain docker, if you are willing to restart the container.  

But I think what we all want is doing it without a restart.

**Comment by ggrrll** [comment]
I would also be interested... 
I guess the answer to my question is currently negative (?)... 
https://stackoverflow.com/questions/71591606/change-existing-labels-of-a-docker-image

**Comment by jasonhildebrand** [comment]
I'd like the ability to be able to add label to existing volumes, in order to control which volumes are included in my offsite backup.  By default I back up all volumes.  But certain volumes do not need to be backed up.  Volume labels would be an elegant solution.

**Comment by Mabenan** [comment]
Is there any news about this would really like to have this to utilize it for traefik

**Comment by XFNeo** [comment]
Up, I need this feature for traefik too

**Comment by gingerbeardman** [comment]
Would be very useful, I just had to rebuild an image just to add a LABEL. Thanks

**Comment by parity3** [comment]
The new docker server 24.0.3 causes some compose files to error with no recourse if certain labels for a network do not match what is expected. I think it's enforcing the same associated project/stack labels, but there could be other labels.

This introduces some extra pain when making incremental service updates; the workaround is to stop all services on the network, remove the network and docker compose up from scratch, possibly introducing more down time for global-style services.

Why have the network labels changed in the first place, you ask? It's because of some refactors in the compose file that have existed for months (where older docker server versions did not care about network labels).

**Comment by akha666** [comment]
> Up, I need this feature for traefik too

+1, Same here. it will be very useful for Traefik to Add/Remove Labels to the running Container

**Comment by rojas-diego** [comment]
Bumping this for the same reasons as most, looking for a way to do on the fly updates to container labels for use with traefik and other solutions that depend on labels. Also have some use cases related to attaching custom metadata to running containers.

**Comment by Clovel** [comment]
+1, for Traefik usage as well !

> Mutable container labels would be _really_ handy. I'm using [traefik](https://traefik.io/), a reverse proxy, which automatically detects docker containers and reads traefik-namespaced container labels for automatic configuration. Setting a label on a running container would allow on-the-fly rule changes.

I came for this, glad to see I'm not the only one with this need.

**Comment by pytimer** [comment]
+1, hope this feature.

I want to avoid kubelet gc  delete some images,  kubelet prevent delete image that has the pinned fields. CRI support `Pinned` field to do it, ref: https://github.com/kubernetes/enhancements/pull/2694, if docker support update the image labels, cri-docker can tell kubelet that image should pinned, but the docker cannot update the label to the image, so this way maybe cannot easy for me.

**Comment by christianh814** [comment]
+1 for this as well 🙏🏽

**Comment by Stocy** [comment]
+1 for this, would be very handy to not have to restart containers to update their labels. 
Use-case here is for Treafik, but I'm sure plenty of other stuff uses container labels...

Happy new year in the meantime !

**Comment by ulope** [comment]
The same use case as traefik is true for the Caddy reverse proxy ([via third party label support](https://github.com/lucaslorentz/caddy-docker-proxy)).

**Comment by davidban77** [comment]
+1, same use case for Traefik/Caddy etc.. Plus custom orchestration of containers based on label data.

**Comment by santhoshpsk** [comment]
I can't believe a feature is being asked to be implemented in docker for 9 years from 2015. Is this still under development?

**Comment by amaczuga** [comment]
Hear, hear. Please consider the feature for upcoming releases. Thank you.

**Comment by sub20hz** [comment]
+1, also using Traefik+Cady but imagine this would be broadly useful across container orchestration

**Comment by ashap5** [comment]
+1 a lot of docker integrations work via labels
Homepage,traefik,authentik , caddy ...

And rn you are forced to have downtime just to change a label

**Comment by Clovel** [comment]
In my team we are abandoning the Traefik Docker Provider because of this. Not using said provider causes problems but not being able to update the labels at container runtime is blocking us from moving forward and causes even more issues. 

Too bad. 

If the dev team reconsiders and implements this we might switch back to the Docker labels & the Docker Provider one day.

**Comment by xoscar** [comment]
+1

**Comment by TheCataliasTNT2k** [comment]
Any update on this?

**Comment by cobolbaby** [comment]
+1
